Introduce `dockerTools.buildNixShellImage` (#172736)
NixOS/nixpkgs: doc/builders/images/dockertools.section.md

```

Creating base files like `/etc/passwd` or `/etc/login.defs` is necessary for shadow-utils to manipulate users and groups.

## buildDerivationImage {#ssec-pkgs-dockerTools-buildDerivationImage}

> Create a Docker image that sets up the shell/build environment of a derivation. With Docker, this environment is similarly sandboxed as if the derivation were built with `nix-build`. But it also behaves like an interactive `nix-shell`. If the derivation is fully buildable (i.e. `nix-build` can be used on it), running `buildDerivation` inside such a Docker container will build the derivation, with all its outputs being available in the correct `/nix/store` paths, pointed to by the respective environment variables like `$out`, etc.

Author: Good point, I'll add that

::: {.warning}
The behavior doesn't match nix-shell exactly and this function may not work correctly for e.g. fixed-output derivations, CA derivations, impure derivations and other special types of derivations.
:::

`drv`

: The derivation whose environment this docker image should be based on.

>     Adding additional packages to the docker image is possible by e.g. extending the list of `buildInputs` of this derivation like

Reviewer: Maybe `nativeBuildInputs` is a better example, as that is the typical attr that should be used in a `nix-shell`. (It has tab completion and strictDeps support)

Of course `buildInputs` is valid for libraries you're building with.

>     Adding additional packages to the docker image is possible by e.g. extending the list of `buildInputs` of this derivation like

Author: Sounds good, using `nativeBuildInputs` instead

    ```nix
    buildDerivationImage {
      drv = someDrv.overrideAttrs (old: {
        nativeBuildInputs = old.nativeBuildInputs or [] ++ [
          somethingExtra
        ];
      });
      # ...
    }
    ```

    Similarly, you can extend the image initialization script by extending `shellHook`

`name` _optional_

: The name of the resulting image.

    *Default:* the `drv`'s name + `"-env"`

`tag` _optional_

: Tag of the generated image.

>     *Default:* the output path's hash

Author: I just copied the description from the other functions in this file. But indeed, it's completely undocumented what output path is used. For `buildImage` it appears to be the actual output path of `buildImage` itself. But for `streamLayeredImage`/`buildLayeredImage` (which this new function is using), the hash comes from a somewhat internal `-conf.json` derivation's output path. I feel like this should be fixed in a separate PR though.

>     *Default:* the output path's hash

Reviewer: You can still clarify that it's an image derivation hash, not the hash of the derivation argument.

>     *Default:* the output path's hash

Author: Did this

  };
}
```

## buildNixShellImage {#ssec-pkgs-dockerTools-buildNixShellImage}

> Create a Docker image that sets up the `nix-shell` environment of a derivation. With Docker, this environment is similarly sandboxed as if the derivation were built with `nix-build`, with a major difference being that access to the internet is allowed. It additionally also behaves like an interactive `nix-shell`, running things like `shellHook` and setting an interactive prompt. If the derivation is fully buildable (i.e. `nix-build` can be used on it), running `buildDerivation` inside such a Docker image will build the derivation, with all its outputs being available in the correct `/nix/store` paths, pointed to by the respective environment variables like `$out`, etc.

Reviewer: _similar - as if_ is a bit too strong. This (like `nix-shell`) is another multi-purpose tool that compromises accuracy for utility. We can strive to make it similar, but succeeding is counter to the other goals.

```suggestion
Create a Docker image that sets up an environment similar to that of running `nix-shell` on a derivation. When run in Docker, this environment somewhat resembles the Nix sandbox typically used by `nix-build`, with a major difference being that access to the internet is allowed. It additionally also behaves like an interactive `nix-shell`, running things like `shellHook` and setting an interactive prompt. If the derivation is fully buildable (i.e. `nix-build` can be used on it), running `buildDerivation` inside such a Docker image will build the derivation, with all its outputs being available in the correct `/nix/store` paths, pointed to by the respective environment variables like `$out`, etc.
```

::: {.warning}

> The behavior doesn't match nix-shell exactly and this function may not work correctly for e.g. fixed-output derivations, CA derivations, impure derivations and other special types of derivations.

Reviewer: ```suggestion
The behavior doesn't match nix-shell or nix-build exactly and this function is known not to work correctly for e.g. fixed-output derivations, CA derivations, impure derivations and other special types of derivations.
```

These have extra behavior that don't make much sense to implement, so we can confidently state that they're broken.

:::

> `drv`

Reviewer 2: It seems we don't have a convention across the manual, but introducing the section would be helpful.
```suggestion
### Arguments

`drv`
```

: The derivation whose environment this Docker image should be based on.

    Adding additional packages to the Docker image is possible by e.g. extending the list of `nativeBuildInputs` of this derivation like

    ```nix
    buildNixShellImage {
      drv = someDrv.overrideAttrs (old: {
        nativeBuildInputs = old.nativeBuildInputs or [] ++ [
          somethingExtra
        ];
      });
      # ...
    }
    ```

    Similarly, you can extend the image initialization script by extending `shellHook`

`name` _optional_

: The name of the resulting image.

>     *Default:* the `drv`'s name + `"-env"`

Reviewer 2: Would it be fair to use Nix language here? You use it below. It's not as clear-cut as `pkgs.something`, but we mention `drv` right next to this.

```suggestion
    *Default:* `drv.name + "-env"`
```

Author: Sounds good!

`tag` _optional_

: Tag of the generated image.

>     *Default:* the resulting image derivation output path's hash

Reviewer 2: This is a handful... Also we seem to be using output path and store path interchangibly. Are they synonymous? If yes, we should use exactly one of them everywhere.
```suggestion
    *Default:* the store path hash of the resulting image's derivation
```

Author: This was also discussed in https://github.com/NixOS/nixpkgs/pull/172736#discussion_r873874868. Notably it's not the store path hash of the resulting image's derivation, but rather some internal `config.json` file.

`uid`/`gid` _optional_

> : The user/group id to run the container as. This is like a nixbld build user

Reviewer 2: ```suggestion
: The user/group ID to run the container as. This is like a `nixbld` build user.
```

    *Default:* 1000/1000

`homeDirectory` _optional_

> : The home directory of the user

Reviewer 2: Which user exactly though?
```suggestion
: The home directory of the user.
```

    *Default:* `/build`

`shell` _optional_

> : The path to the bash binary to use as the shell. This shell is started when running the image

Reviewer 2: ```suggestion
: The path to `bash` to be used as the shell. This shell is started when running the image.
```

Author: I don't think this is right, did you mean to suggest

> The path to the `bash` binary to be used as the shell. This shell is started when running the image.

?

Reviewer 2: Updated the suggestion.

    *Default:* `pkgs.bashInteractive + "/bin/bash"`

`command` _optional_

> : Run this command in the environment of the derivation, in an interactive shell. See `--command` in `man nix-shell`

Reviewer 2: Why don't we link to the Nix manual here?

```suggestion
: Run this command in the environment of the derivation, in an interactive shell. See `--command` in `man nix-shell`.
```

Author: Good idea, doing that

    *Default:* (none)

`run` _optional_

> : Same as `command`, but runs the command in a non-interactive shell instead. See `--run` in `man nix-shell`

Reviewer 2: Why don't we link to the Nix manual here?
```suggestion
: Same as `command`, but runs the command in a non-interactive shell instead. See `--run` in `man nix-shell`.
```

    *Default:* (none)

> The following shows how to build the `pkgs.hello` package inside a Docker container built with `buildNixShellImage`.

Reviewer 2: ```suggestion
### Example

The following shows how to build the `pkgs.hello` package inside a Docker container built with `buildNixShellImage`.
```

```nix
with import <nixpkgs> {};
dockerTools.buildNixShellImage {
  drv = hello;
}
```

> Now to build, load and run the image:
> 
> ```
> $ nix-build hello.nix
> these 8 derivations will be built:
>   /nix/store/xmw3a5ln29rdalavcxk1w3m4zb2n7kk6-nix-shell-rc.drv
> [...]
> Creating layer 56 from paths: ['/nix/store/crpnj8ssz0va2q0p5ibv9i6k6n52gcya-stdenv-linux']
> Creating layer 57 with customisation...
> Adding manifests...
> Done.
> /nix/store/cpyn1lc897ghx0rhr2xy49jvyn52bazv-hello-2.12-env.tar.gz
> 
> $ docker load -i result
> 0d9f4c4cd109: Loading layer [==================================================>]   2.56MB/2.56MB
> [...]
> ab1d897c0697: Loading layer [==================================================>]  10.24kB/10.24kB
> Loaded image: hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
> 
> $ docker run -it hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
> 
> [nix-shell:/build]$ buildDerivation
> unpacking sources
> unpacking source archive /nix/store/8nqv6kshb3vs5q5bs2k600xpj5bkavkc-hello-2.12.tar.gz
> [...]
> patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
> checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...
> 
> [nix-shell:/build]$ $out/bin/hello
> Hello, world!
> ```

Reviewer 2: ````suggestion
Build the derivation:

```console
nix-build hello.nix
```

    these 8 derivations will be built:
      /nix/store/xmw3a5ln29rdalavcxk1w3m4zb2n7kk6-nix-shell-rc.drv
    ...
    Creating layer 56 from paths: ['/nix/store/crpnj8ssz0va2q0p5ibv9i6k6n52gcya-stdenv-linux']
    Creating layer 57 with customisation...
    Adding manifests...
    Done.
    /nix/store/cpyn1lc897ghx0rhr2xy49jvyn52bazv-hello-2.12-env.tar.gz

Load the image:

```console
docker load -i result
```

    0d9f4c4cd109: Loading layer [==================================================>]   2.56MB/2.56MB
    ...
    ab1d897c0697: Loading layer [==================================================>]  10.24kB/10.24kB
    Loaded image: hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz

Run the container:

```console
docker run -it hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
```

    [nix-shell:/build]$

In the running container, run the build:

```console
buildDerivation
```

    unpacking sources
    unpacking source archive /nix/store/8nqv6kshb3vs5q5bs2k600xpj5bkavkc-hello-2.12.tar.gz
    ...
    patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
    checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...

Check the build result:

```console
$out/bin/hello
```

    Hello, world!
```
````

      /nix/store/xmw3a5ln29rdalavcxk1w3m4zb2n7kk6-nix-shell-rc.drv
    ...
    Creating layer 56 from paths: ['/nix/store/crpnj8ssz0va2q0p5ibv9i6k6n52gcya-stdenv-linux']
    Creating layer 57 with customisation...
    Adding manifests...
    Done.
    /nix/store/cpyn1lc897ghx0rhr2xy49jvyn52bazv-hello-2.12-env.tar.gz

Load the image:
```console
docker load -i result
```

    0d9f4c4cd109: Loading layer [==================================================>]   2.56MB/2.56MB
    ...
    ab1d897c0697: Loading layer [==================================================>]  10.24kB/10.24kB
    Loaded image: hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz

Run the container:
```console
docker run -it hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
```

    [nix-shell:/build]$

In the running container, run the build:
```console
buildDerivation
```

    unpacking sources
    unpacking source archive /nix/store/8nqv6kshb3vs5q5bs2k600xpj5bkavkc-hello-2.12.tar.gz
    ...
    patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
    checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...

Check the build result:

> ```console

Reviewer 2: I wonder why people don't keep an empty line between paragraphs and code blocks. @olaflkingt I also see this in your PRs. My mental markdown parser trips over this. I won't annoy people about this (yet), but maybe someone can explain what that's about. IMHO block-level elements should be separated by empty lines. I know this is not how all the Markdowns out there see things, but we write this stuff for people to read and only very rarely for machines to process.

Reviewer 3: The answer is: small screens. I started to clean that up. iknowsomepeoplethingwhitespacesandlinebreaksimprovereadability.

Author: I didn't think about this a lot, but a small part of my brain said to not put what is not needed, and there's no spacing needed here. Contrary to text paragraphs where you do need an extra newline. But I really don't care, I might as well add another newline for consistency :)

Author: Did that in a new force-push
